README.md
# ImageArtist 1.0.0

ImageArtist is a php gd Wrapper which makes Image manipulation insanely easy, we introduce ImageArtist as Photoshop for php 

This project is an initiative of [Treinetic (Pvt) Ltd](http://www.treinetic.com), Sri Lanka.


![Issues](https://img.shields.io/github/issues/Treinetic/ImageArtist.svg)
[![Software License](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)
![Forks](https://img.shields.io/github/forks/Treinetic/ImageArtist.svg)
![Stars](https://img.shields.io/github/stars/Treinetic/ImageArtist.svg)
![Twitter](https://img.shields.io/twitter/url/https/github.com/Treinetic/ImageArtist.svg?style=social)


## Requirements

PHP 5.0 >

**Optional** ( Consider installing imagick along with pango for complex unicode Texts )
- [Imageick](http://php.net/manual/en/book.imagick.php) with [Pango](http://www.pango.org/)

## Install

The package can be installed via composer:
``` bash
$ composer require treinetic/imageartist
```

## Usage ( Let's Dive In )

### Very Basics

* Load Image, Most of the Classes Extends the Image class so if you know how to load an Image then you know how to use most of the library classes

```php
$image = new Image("./morning.jpeg");
```
**Note :** Remember to import ( use ) the class first, see demo folder for more complete code.

* Image Resizing

```php
$image->scale(40); //scales the image 40%
$image->scaleToHeight(1024); //scales the image keeping height 1024
$image->scaleToWidth(1024); //scales the image keeping width 1024
$image->resize(1024,768); //resizes the image to a given size 

```

* Image Attributes

```php
$new_wdith = $image->getWidth();
$new_height = $image->getHeight();
```

* Change Format/ Save to Disk
```php
$image->convertTo(IMAGETYPE_PNG);
$image->save("./newImage.jpg",IMAGETYPE_PNG);
```

* Orientation
```php
$image->rotate(50); 
$image->flipV(); // Vertical Flip
$image->flipH(); // Horizontal Flip
```


* Other Usefull Operations
```php
$base64URL = $image->getDataURI();
/* 
    Image class will return itself for following methods but in Shape classes it will be a new Image 
    to keep the idea of Shape Consistant
 */
$image->crop(100,100,300,300);
$image->merge(new Image("./city.jpg"),10,10);
```

### Awesome Stuff
**Lets do some things that matter**

* **Create a shape**, for creating a custom shape you can directly use Either `CircularShape.php` or `PolygonShape.php` however if you are looking for a standard shape which is not yet existing then still you can use either `CircularShape.php` or `PolygonShape.php` to create them but we encorage you to contribute to the project by adding those shape
* Lets Create a **Triangle** ( There are few other Pre made shapes for your use )

```php
$triangle = new Triangle("./city.jpg");
$triangle->scale(60);
$triangle->setPointA(20,20,true); //setting point A, i'm using preentages but you can use px as well
$triangle->setPointB(80,20,true);
$triangle->setPointC(50,80,true);
$triangle->build();
```

<p align="center">
  <img width="400" src="https://raw.githubusercontent.com/Treinetic/ImageArtist/images/img/triangle.png"/>
</p>

* Create **Create Custom Shape**, lets see how we can make a nice Pentagon

```php
$pentagon = new PolygonShape("./morning.jpeg");
$pentagon->scale(60);
$pentagon->push(new Node(50,0, Node::$PERCENTAGE_METRICS));
$pentagon->push(new Node(75,50, Node::$PERCENTAGE_METRICS));
$pentagon->push(new Node(62.5,100, Node::$PERCENTAGE_METRICS));
$pentagon->push(new Node(37.5,100, Node::$PERCENTAGE_METRICS));
$pentagon->push(new Node(25,50, Node::$PERCENTAGE_METRICS));
$pentagon->build();
```
<p align="center">
<img width="400" src="https://raw.githubusercontent.com/Treinetic/ImageArtist/images/img/polygon.png"/>
</p>

* Lets **Merge Two Triangles** 

```php
$tr1 = new Triangle("./city.jpg");
$tr1->scale(60);
$tr1->setPointA(0,0,true);
$tr1->setPointB(100,0,true);
$tr1->setPointC(100,100,true);
$tr1->build();

$tr2 = new Triangle("./morning.jpeg");
$tr2->scale(60);
$tr2->setPointA(0,0,true);
$tr2->setPointB(0,100,true);
$tr2->setPointC(100,100,true);
$tr2->build();

$tr1->resize($tr1->getWidth(),$tr2->getHeight());

$img = $tr1->merge($tr2,0,0);
$img->scale(70);
```

<p align="center">
<img src="https://raw.githubusercontent.com/Treinetic/ImageArtist/images/img/merge.png"/>
</p>

* Let's Do Some Photoshop Shall we ? **Create a Facebook cover**
 
```php
/* Let's add an overlay to this */
$overlay = new Overlay($img->getWidth(),$img->getHeight(),new Color(0,0,0,80));
$img->merge($overlay,0,0);
/* hmmm.. lets add a photo */
$circle = new CircularShape("./person.jpg");
$circle->build();
$img = $img->merge($circle,($img->getWidth()-$circle->getWidth())/2,($img->getHeight() - $circle->getHeight())/2);
/* I think I should add some Text */
$textBox = new TextBox(310,40);
$textBox->setColor(Color::getColor(Color::$WHITE));
$textBox->setFont(Font::getFont(Font::$NOTOSERIF_REGULAR));
$textBox->setSize(20);
$textBox->setMargin(2);
$textBox->setText("We Are Team Treinetic");
$img->setTextBox($textBox,($img->getWidth()-$textBox->getWidth())/2,$img->getHeight()* (5/7));

$img->dump(); //development purposes only
```

<p align="center">
<img src="https://raw.githubusercontent.com/Treinetic/ImageArtist/images/img/cover.png"/>
</p>


## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

- [Imal Hasaranga Perera](https://github.com/imalhasaranga)
- [Nuwan Chamara](https://github.com/nuwanchamara)
- [All Contributors](../../contributors)


## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.

CHANGELOG.md
# Changelog
# 1.0.0 (Ready For The World) - 2017-09-24
- We feel we are stable enough go live.

# 0.3.1 (beta) Released - 2017-09-24
- Bug fixed on Text angles, both in Imagick Writing Strategy as well as GDWritingStrategy
- More minor improvements, I think we are ready for the first release version of ImageArtist :) 

# 0.3.0 (alpha) Released - 2017-09-13
- TextWriting Strategy was previously switched automatically based on imageick and pango avialability, with this version we are providing another method called 'WriteFactory.overrideWriteStrategySelection()' to manually switch between writing strategies
- 'flipH()', 'flipV()', 'rotate($degrees)' methods added to 'Image.php'
- For fixing Orientation based on exif value 'OrientationFixer.php' helper added

# 0.2.1 patch Released - 2017-09-12 
there was a bug in the `Rectangle.php` for the older php versions


# 0.2.0 (alpha) Released - 2017-09-12
- method `getBase64URL` renamed to `getDataURI` in `Image.php` 
- strategy pattern implemented for textWriting to decide which writer ( gd or imagick ) to be selected, we decided to go for this to support complex unicode rendering

# 0.1.0-alpha Released - 2017-09-10
0.1.0 alpha release with lot of code optimizations and more hidden bug fixes, this include 
- Image merge function is now supoorting negative cordinates
- PolygonShape.php build method was not shrinking to the shape size bug fixed.
- Rectangle.php helper was added wrapping some linear algebric logic to support image croping and boundary calculations
- Square Shape added ( but it may get renamed soon )


# 0.1.0-alpha-rc1
0.1.0 alpha candidate is now progressing, more focused on code stability and build more functionality ontop of the current core libraries

# 0.0.1-alpha - 2017-09-08
- Contriubtion team did the first release of ImageArtist for internal use of the company project, this first release is having following features
### Features
- Polygon and the CircleShape aded 
- TextBox for Text writing on top of Images
- Overlay creationg
- Powerfull class for basic image operations like resizing, cropping, saving to differnt format etc


# First Breath (Repo Created)
- Imal hasaranga perera CEO of Treinetic (pvt) took the first step of writing an interactive easy to use image manipulation library after understaning the waste of time and huge number of code lines that developers has to write just to get done a simple image manipulation task done



demo/awesome.php
<?php
/**
 * Created by PhpStorm.
 * User: Nuwan
 * Date: 08/09/2017
 * Time: 14:42
 */


use \Treinetic\ImageArtist\lib\Shapes\Triangle;
use \Treinetic\ImageArtist\lib\Shapes\PolygonShape;
use \Treinetic\ImageArtist\lib\Commons\Node;
use \Treinetic\ImageArtist\lib\Text\Color;
use \Treinetic\ImageArtist\lib\Overlays\Overlay;
use \Treinetic\ImageArtist\lib\Shapes\CircularShape;
use \Treinetic\ImageArtist\lib\Text\TextBox;
use \Treinetic\ImageArtist\lib\Text\Font;

require('../vendor/autoload.php');

/** @var \Treinetic\ImageArtist\lib\Image $img */


$triangle = new Triangle("./city.jpg");
$triangle->scale(60);
$triangle->setPointA(20,20,true); //setting point A, i'm using preentages but you can use px as well
$triangle->setPointB(80,20,true);
$triangle->setPointC(50,80,true);
$triangle->build();

$triangle->dump();

/*
 * Pentagon
 * */

$pentagon = new PolygonShape("./morning.jpeg");
$pentagon->scale(60);
$pentagon->push(new Node(50,0, Node::$PERCENTAGE_METRICS));
$pentagon->push(new Node(75,50, Node::$PERCENTAGE_METRICS));
$pentagon->push(new Node(62.5,100, Node::$PERCENTAGE_METRICS));
$pentagon->push(new Node(37.5,100, Node::$PERCENTAGE_METRICS));
$pentagon->push(new Node(25,50, Node::$PERCENTAGE_METRICS));
$pentagon->build();

$pentagon->dump();


/*
 * Merge images
 * */

$tr1 = new Triangle("./city.jpg");
$tr1->scale(60);
$tr1->setPointA(0,0,true);
$tr1->setPointB(100,0,true);
$tr1->setPointC(100,100,true);
$tr1->build();

$tr2 = new Triangle("./morning.jpeg");
$tr2->scale(60);
$tr2->setPointA(0,0,true);
$tr2->setPointB(0,100,true);
$tr2->setPointC(100,100,true);
$tr2->build();

$tr1->resize($tr1->getWidth(),$tr2->getHeight());

$img = $tr1->merge($tr2,0,0);
$img->scale(70);

$img->dump();


/* Let's add an overlay to this */
$overlay = new Overlay($img->getWidth(),$img->getHeight(),new Color(0,0,0,80));
$img->merge($overlay,0,0);
/* hmmm.. lets add a photo */
$circle = new CircularShape("./person.jpg");
$circle->build();
$img = $img->merge($circle,($img->getWidth()-$circle->getWidth())/2,($img->getHeight() - $circle->getHeight())/2);
/* I think I should add some Text */
$textBox = new TextBox(310,40);
$textBox->setColor(Color::getColor(Color::$WHITE));
$textBox->setFont(Font::getFont(Font::$NOTOSERIF_REGULAR));
$textBox->setSize(20);
$textBox->setMargin(2);
$textBox->setText("We Are Team Treinetic");
$img->setTextBox($textBox,($img->getWidth()-$textBox->getWidth())/2,$img->getHeight()* (5/7));

$img->dump();